### PR TITLE
fix: derrive min sdk values from android lint

### PR DESCRIPTION
**Pull request by jamesonwilliams**
For the purpose of re-enabling NewApi and other API-related Android Lint
warnings, all `abortOnError` options are removed. In response, some
unrelated lint warnings are being addressed or suppressed in code, as a
result of this commit.

To derive an appropriate `minSdkVersion` for each module, `minSdkVersion`
was stripped from all modules, and then updated according to the project
build requirements.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: aws-android-sdk-core/src/main/java/com/amazonaws/internal/keyvaluestore/AWSKeyValueStore.java [hunk 1]
```diff
@@ -471,11 +464,11 @@ private AlgorithmParameterSpec getInitializationVector(final String keyOfDataInS
     }
 
     private AlgorithmParameterSpec getAlgorithmParameterSpecForIV(byte[] iv) {
-        return
-            //@apiLevel23Start
-            apiLevel >= ANDROID_API_LEVEL_23 ? new GCMParameterSpec(CIPHER_AES_GCM_NOPADDING_TAG_LENGTH_LENGTH_IN_BITS, iv) :
-            //@apiLevel23End
-            new IvParameterSpec(iv);
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
```
Comment (jamesonwilliams) [inline aws-android-sdk-core/src/main/java/com/amazonaws/internal/keyvaluestore/AWSKeyValueStore.java:467]: Android Lint can't figure out that this code is correctly guarded, unless these standard `Build` values are used.

File: aws-android-sdk-kinesisvideo/src/main/java/com/amazonaws/kinesisvideo/encoding/ChunkDecoder.java [hunk 2]
```diff
@@ -53,7 +53,7 @@ private ChunkDecoder() {
 
         try {
             final String headersAsString =
-                    readInputStream(inputStream, PAYLOAD_DELIMITER.getBytes(StandardCharsets.UTF_8));
```
Comment (jamesonwilliams) [inline aws-android-sdk-kinesisvideo/src/main/java/com/amazonaws/kinesisvideo/encoding/ChunkDecoder.java:56]: `StandardCharsets` bumps the `minSdkVersion` up to 19, for no good reason. Don't use it.

File: aws-android-sdk-core/src/main/java/com/amazonaws/internal/keyvaluestore/KeyNotGeneratedException.java [hunk 3]
```diff
@@ -16,8 +16,5 @@ public KeyNotGeneratedException(String s, Throwable throwable) {
     public KeyNotGeneratedException(Throwable throwable) {
         super(throwable);
     }
-
-    protected KeyNotGeneratedException(String s, Throwable throwable, boolean b, boolean b1) {
-        super(s, throwable, b, b1);
```
Comment (jamesonwilliams) [inline aws-android-sdk-core/src/main/java/com/amazonaws/internal/keyvaluestore/KeyNotGeneratedException.java:21]: This was not being used, and it has an insanely high API level requirement. Get rid of it!

File: aws-android-sdk-kinesisvideo-archivedmedia/build.gradle [hunk 4]
```diff
@@ -4,16 +4,12 @@ android {
     compileSdkVersion 28
 
     defaultConfig {
-        minSdkVersion 15
+        minSdkVersion 21 // for kinesisvideo, which uses camera2 APIs.
```
Comment (TrekSoft) [inline aws-android-sdk-kinesisvideo-archivedmedia/build.gradle:7]: Hmm, will this have an impact on our existing customers? What was happening currently with the minSdkVersion being too low? Is there only a limited number of methods needing this we could wrap with RequiresAPI annotations?

File: aws-android-sdk-kinesisvideo/build.gradle [hunk 5]
```diff
@@ -5,7 +5,7 @@ android {
     useLibrary 'org.apache.http.legacy'
 
     defaultConfig {
-        minSdkVersion 15
+        minSdkVersion 21 // android.hardware.camera2
```
Comment (TrekSoft) [inline aws-android-sdk-kinesisvideo/build.gradle:8]: Same question here as before

File: aws-android-sdk-mobile-client/src/main/AndroidManifest.xml [hunk 6]
```diff
@@ -1,9 +1,17 @@
...
            xmlns:amazon="http://schemas.amazon.com/apk/res/android"
+           xmlns:tools="http://schemas.android.com/tools"
            package="com.amazonaws.mobile.client">
 
     <uses-permission android:name="android.permission.INTERNET"/>
     <uses-permission android:name="android.permission.ACCESS_NETWORK_STATE"/>
- 
+
+    <!--
+         The hostest UI requires API 16, for Custom Tabs support in the
```
Comment (TrekSoft) [inline aws-android-sdk-mobile-client/src/main/AndroidManifest.xml:11]: *hosted UI

File: aws-android-sdk-kinesisvideo-archivedmedia/build.gradle (hunk 4 above)
Comment (jamesonwilliams) [reply aws-android-sdk-kinesisvideo-archivedmedia/build.gradle:7]: Previously, the `lintOptions { abortOnError false }` was suppressing the build error about the `minSdkVersion` being too low.

The API level is ultimately coming from a few files in the `-kinesisvideo` package, which then bump up this package. Let me see if I can use the `RequiresVersion`, to lower requirements on the two packages.

File: aws-android-sdk-mobile-client/src/main/AndroidManifest.xml (hunk 6 above)
Comment (jamesonwilliams) [reply aws-android-sdk-mobile-client/src/main/AndroidManifest.xml:11]: Haha! Oops. Will fix.

File: aws-android-sdk-kinesisvideo-archivedmedia/build.gradle (hunk 4 above)
Comment (jamesonwilliams) [reply aws-android-sdk-kinesisvideo-archivedmedia/build.gradle:7]: I started going down this path -- it is ultimately a lot of files that raise that API level. I think it'd be best to do it globally on the module for now. If people complain and need lower API level support, we could re-evaluate in the future.

**Pull request merged**